zkevm-circuits/src/ecc_circuit.rs

```diff
@@ -674,43 +655,134 @@ impl<F: Field, const XI_0: i64> EccCircuit<F, XI_0> {
+                    ec_add_decomposed.point_q.y_cells.clone(),
+                    keccak_powers.iter().cloned(),
+                ),
+            },
+            point_r: G1Assigned {
+                x_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                    ctx,
+                    ec_add_decomposed.point_r.x_cells.clone(),
+                    keccak_powers.iter().cloned(),
+                ),
+                y_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                    ctx,
+                    ec_add_decomposed.point_r.y_cells.clone(),
+                    keccak_powers.iter().cloned(),
+                ),
+            },
+        }
+    }
+
+    /// Handles Phase2 for EcMul operation and returns the RLC'd x and y co-ordinates of the G1
+    /// elements, and the assigned scalar field element.
+    fn assign_ec_mul(
+        &self,
+        ctx: &mut Context<F>,
+        ec_mul_decomposed: &EcMulDecomposed<F>,
+        ecc_chip: &EccChip<F, FpConfig<F, Fq>>,
+        keccak_powers: &[QuantumCell<F>],
+    ) -> EcMulAssigned<F> {
+        EcMulAssigned {
+            point_p: G1Assigned {
+                x_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                    ctx,
+                    ec_mul_decomposed.point_p.x_cells.clone(),
+                    keccak_powers.iter().cloned(),
+                ),
+                y_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                    ctx,
+                    ec_mul_decomposed.point_p.y_cells.clone(),
+                    keccak_powers.iter().cloned(),
+                ),
+            },
+            scalar_s: ec_mul_decomposed.scalar_s.clone(),
+            point_r: G1Assigned {
+                x_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                    ctx,
+                    ec_mul_decomposed.point_r.x_cells.clone(),
+                    keccak_powers.iter().cloned(),
+                ),
+                y_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                    ctx,
+                    ec_mul_decomposed.point_r.y_cells.clone(),
+                    keccak_powers.iter().cloned(),
+                ),
+            },
+        }
+    }
+
+    /// Handles Phase2 for EcPairing operation and returns the RLC'd input bytes.
+    fn assign_ec_pairing(
+        &self,
+        ctx: &mut Context<F>,
+        ec_pairing_decomposed: &EcPairingDecomposed<F>,
+        ecc_chip: &EccChip<F, FpConfig<F, Fq>>,
+        keccak_powers: &[QuantumCell<F>],
+    ) -> EcPairingAssigned<F> {
+        EcPairingAssigned {
+            input_rlc: ecc_chip.field_chip().range().gate().inner_product(
+                ctx,
+                ec_pairing_decomposed.input_cells.clone().into_iter().rev(),
+                keccak_powers.iter().cloned(),
+            ),
+            success: ec_pairing_decomposed.success,
+        }
+    }
+
+    /// Assign the G1 EcPoint and return its decomposed state.
```

Comment:
IIUC, this function asserts that the statement
```
ec_point is a valid crt form of g1
```
but it does not assert
```
(x_cells, y_cells) are a valid byte representation of g1
```

is the second statement constraint somewhere else?

Reply:
This constraint is missing in `EccCircuit`. I will add it now for every relevant field.